### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by Evgenia789**
Task 1

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -133,4 +133,4 @@ service_account.json
 .pyre/
```
Comment (Phaere92) [inline .gitignore:133]: 1

File: src/tgbot_expenses/config.py [hunk 2]
```diff
@@ -38,22 +38,30 @@ class AllowedIds:
...
-    currency: str
-    total_amount: str
+    postgres_host: str
+    postgres_port: str
+    postgres_user: str
+    postgres_password: str
+    postgres_db: str
+    db_url: str = None
+
+    def __post_init__(self):
```
Comment (atsanda) [inline src/tgbot_expenses/config.py]: What if users passes url, but it is not consistent with other variables like `postgres_user` or `postgres_password`. If you can automatically derive the url from the arguments, then it'd better to create it as a property method @property which calculates url on the fly. If you use `db_url` extensively, use can leave it like this, but disable an option to pass `db_url` as an argument to avoid inconsistencies.

File: src/tgbot_expenses/config.py [hunk 3]
```diff
@@ -38,22 +38,30 @@ class AllowedIds:
 
 
 @dataclass
-class GoogleTables:
+class PostgresDB:
```
Comment (atsanda) [inline src/tgbot_expenses/config.py]: PostgresDBConfig? It is not the Postgres db itself, right?

File: src/tgbot_expenses/config.py [hunk 4]
```diff
@@ -99,11 +107,12 @@ def load_config(path: str) -> Config:
             id_1=ids.getint("ID_1"),
             id_2=ids.getint("ID_2")
         ),
-        googletables=GoogleTables(
-            spreadsheet=googletables.get("spreadsheet"),
-            expenses=googletables.get("expenses"),
-            incomes=googletables.get("incomes"),
-            currency=googletables.get("currency"),
-            total_amount=googletables.get("total_amount")
+        postgres_db=PostgresDB(
```
Comment (atsanda) [inline src/tgbot_expenses/config.py]: Dataclasses can be initialized directly from dictionaries. 
```python
PostgresDB(**postgres_db)
```
Please, rewrite it so it works this way.

File: src/tgbot_expenses/config.py [hunk 5]
```diff
@@ -64,11 +72,11 @@ class Config:
...
-        googletables (GoogleTables): Google Sheets tables configuration.
+        postgres_db (PostgresDB): PostgreSQL database configuration.
     """
     tg_bot: TgBot
     ids: AllowedIds
-    googletables: GoogleTables
+    postgres_db: PostgresDB
 
 
 def load_config(path: str) -> Config:
```
Comment (atsanda) [inline src/tgbot_expenses/config.py]: It seems like `load_config` can be implemented as a class method for the `Config` class. What's the reason to put it aside?

File: src/tgbot_expenses/constants.py [hunk 6]
```diff
@@ -8,21 +8,24 @@ class QuestionText:
     """
```
Comment (atsanda) [inline src/tgbot_expenses/constants.py:8]: Why do you use dataclasses here? Do you override these constants down the road? If not, then you can use the as module variables, without wrapping into a dataclass.

File: src/tgbot_expenses/dialogs/callbacks/expenses/accounts.py [hunk 7]
```diff
@@ -6,13 +6,13 @@
 from src.tgbot_expenses.states.chat_states import StateChat
 
 
-@Bot.callback_query_handler(state=StateChat.Bill)
-async def callbacks_get_bill(query: types.CallbackQuery,
-                             state: FSMContext) -> None:
+@Bot.callback_query_handler(state=StateChat.Account)
```
Comment (atsanda) [inline src/tgbot_expenses/dialogs/callbacks/expenses/accounts.py:9]: This change from Bill -> Account affected too many files. It'd be better to separate changes. Changes like this one should be in another branch with its own PR. For now it is ok, though.

File: src/tgbot_expenses/middlewares/unknown.py
Comment (atsanda) [inline src/tgbot_expenses/middlewares/unknown.py:1]: unknown seems to be a bad name. It looks like a default name for a file when it was created. Looks confusing to me.

File: src/tgbot_expenses/models/expense_tracking_models.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        monthly_limit (Decimal): maximum amount that can be spent in this
+                                 category per month
+        category_status (str): status of the category, defaults to "active"
+
+    Relationships:
+        expenses (List[Expense]): all expenses that belong to this category
+    """
+    __tablename__ = 'categories'
+
+    id = Column(Integer, primary_key=True, index=True,  autoincrement=True)
```
Comment (atsanda) [inline src/tgbot_expenses/models/expense_tracking_models.py]: For simple names like "Integer" it is a bit better not import them using "from <> import Integer", but rather define a shortcut "import long_name as lng" and then use "lng.Integer" This way you can be sure that another import is not overriding it, and it is also clear where it comes from

File: src/tgbot_expenses/models/expense_tracking_models.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    name = Column(String, nullable=False)
```
Comment (atsanda) [inline src/tgbot_expenses/models/expense_tracking_models.py]: Please, provide `length` inside String too. This is valid for almost all yor String-s, unless there is a field with long text when you don't really know the length.

File: src/tgbot_expenses/utils/load_modules.py [hunk 10]
```diff
@@ -1,7 +1,8 @@
+import asyncio
 import os
 
 
-def load_module(name: str, cur_dir: str) -> None:
+async def load_module(name: str, cur_dir: str) -> None:
```
Comment (atsanda) [inline src/tgbot_expenses/utils/load_modules.py:8]: Do you really need this one? Why do you use it. Implicit imports are bad. If you have a bug, it's hard to find then

File: src/tgbot_expenses/database/db.py [hunk 11]
```diff
@@ -1,389 +1,364 @@
...
+
+    async def __aexit__(self, exc_type, exc_val, traceback):
+        """
+        Rolls back any uncommitted changes when the 'async with' block
+        is exited.
+        """
+        await super().__aexit__(exc_type, exc_val, traceback)
+
+
+class AsyncPostgresDB:
```
Comment (atsanda) [inline src/tgbot_expenses/database/db.py:31]: It is not extensible. As your bot grows, this class will have more and more methods for ALL models. It will result in an unreadable and long class. Think about how you can split it. If it is a simple method, like getting a user account by id, may be, you can create a class method for the account model. Or you can introduce a concept of services and put queries there as functions. 
```python
- accounts
   - __init__.py
   - models.py
   - services.py
- expenses
    - __init__.py
    - models.py
    - services.py
 ```
You are free to come up with your own structure, but it should be more flexible and readable. You cannot put all your queries in one class.

File: src/tgbot_expenses/database/db.py [hunk 12]
```diff
@@ -1,389 +1,364 @@
...
-            title=self.config.googletables.total_amount, row=id,
-            column=3, value=float(last_amount)
-        )
-
+        async with AsyncSessionWithEnter(self.engine) as session:
+            account_obj_from = await session.execute(select(Account).where(
+                Account.name == account_from
+            ))
+            account_from = account_obj_from.scalars().first()
+            account_from.balance = account_from.balance - amount_old_currency
```
Comment (atsanda) [inline src/tgbot_expenses/database/db.py]: Is it one transaction? Do you block these rows for update? Otherwise the information introduced between you read `account_from.balance` and write a new balance will be lost.

File: src/tgbot_expenses/dialogs/callbacks/settings/account_settings/account_for_deleting.py [hunk 13]
```diff
@@ -23,13 +24,14 @@ async def callbacks_get_bill_for_deletting(query: types.CallbackQuery,
     """
     await query.message.delete()
 
-    await StateSettings.DeleteBill.set()
+    await StateSettings.DeleteAccount.set()
 
+    accounts = await database.get_all_accounts()
```
Comment (atsanda) [inline src/tgbot_expenses/dialogs/callbacks/settings/account_settings/account_for_deleting.py]: What if database query fails here? Internet connection is broken for example.

**Review by atsanda — CHANGES_REQUESTED** [review]

File: src/tgbot_expenses/config.py (hunk 2 above)
Comment (Evgenia789) [reply src/tgbot_expenses/config.py]: Added property method.

File: src/tgbot_expenses/config.py (hunk 3 above)
Comment (Evgenia789) [reply src/tgbot_expenses/config.py]: Renamed the class name.

File: src/tgbot_expenses/config.py (hunk 4 above)
Comment (Evgenia789) [reply src/tgbot_expenses/config.py]: Rewrote it.

File: src/tgbot_expenses/config.py (hunk 5 above)
Comment (Evgenia789) [reply src/tgbot_expenses/config.py]: Put it inside the class.

File: src/tgbot_expenses/constants.py (hunk 6 above)
Comment (Evgenia789) [reply src/tgbot_expenses/constants.py:8]: I found that this is one of the possible options for storing constants. I added a `frozen` argument that allows to mark data class as immutable. Initially, I planned that I would have two different classes for storing constants to structure the storage.

File: src/tgbot_expenses/dialogs/callbacks/expenses/accounts.py (hunk 7 above)
Comment (Evgenia789) [reply src/tgbot_expenses/dialogs/callbacks/expenses/accounts.py:9]: I understood

File: src/tgbot_expenses/middlewares/unknown.py
Comment (Evgenia789) [reply src/tgbot_expenses/middlewares/unknown.py:1]: Renamed

File: src/tgbot_expenses/models/expense_tracking_models.py (hunk 8 above)
Comment (Evgenia789) [reply src/tgbot_expenses/models/expense_tracking_models.py]: Fixed it

File: src/tgbot_expenses/models/expense_tracking_models.py (hunk 9 above)
Comment (Evgenia789) [reply src/tgbot_expenses/models/expense_tracking_models.py]: Added.

File: src/tgbot_expenses/utils/load_modules.py (hunk 10 above)
Comment (Evgenia789) [reply src/tgbot_expenses/utils/load_modules.py:8]: Yes, because I have a lot of files, and I will have a lot of imports... I have added logging to that function that I can see error.

File: src/tgbot_expenses/database/db.py (hunk 11 above)
Comment (Evgenia789) [reply src/tgbot_expenses/database/db.py:31]: I created `service` directory and service.py for each model.

File: src/tgbot_expenses/database/db.py (hunk 12 above)
Comment (Evgenia789) [reply src/tgbot_expenses/database/db.py]: I changed it.

File: src/tgbot_expenses/dialogs/callbacks/settings/account_settings/account_for_deleting.py (hunk 13 above)
Comment (Evgenia789) [reply src/tgbot_expenses/dialogs/callbacks/settings/account_settings/account_for_deleting.py]: I added exceptions handling.

File: src/tgbot_expenses/dialogs/callbacks/settings/limit_settings/changing_limit.py [hunk 14]
```diff
@@ -25,11 +26,12 @@ async def callbacks_change_limit(query: types.CallbackQuery,
 
     await StateSettings.ChangeLimit.set()
 
+    categories = await get_all_categories_with_retry()
     await Bot.answer(
         message=query.message,
         text=QuestionText.limits,
         reply_markup=get_keyboard_question(
-            button_names=(database.get_all_categories()),
+            button_names=(categories),
```
Comment (atsanda) [inline src/tgbot_expenses/dialogs/callbacks/settings/limit_settings/changing_limit.py]: () are not necessary here

File: src/tgbot_expenses/utils/load_modules.py (hunk 10 above)
Comment (atsanda) [reply src/tgbot_expenses/utils/load_modules.py:8]: it is better to use `from dialogs import *` It is still bad, but less implicit. You can leave it as it is since this feature is a about database, but I'd rethink this part, if I wanted to develop it further

File: src/tgbot_expenses/services/account_service.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+                           to insert.
+    :type account_amount: Decimal
+    :return: None
+    """
+    async with AsyncSessionWithEnter(database.engine) as session:
+        account = Account(name=account_name, balance=account_amount)
+        session.add(account)
+        await session.commit()
+
+    return
```
Comment (atsanda) [inline src/tgbot_expenses/services/account_service.py]: There is no need for return here

File: src/tgbot_expenses/services/account_service.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    """
+    Retrieve the amount associated with the given account ID from
+    the 'accounts' table.
+
+    :param account_id: The ID of the account to retrieve the amount for.
+    :type account_id: int
+    :return: The balance associated with the account.
+    """
+    async with AsyncSessionWithEnter(database.engine) as session:
+        account_obj = await session.execute(select(Account).where(
```
Comment (atsanda) [inline src/tgbot_expenses/services/account_service.py]: Please, select retrieve balance directly within one query. There is no need to select full info of an account. Currently your SQL should look like this
```sql
SELECT * 
FROM Accounts
WHERE id == account_id
```
But you need
```sql
SELECT balance
FROM Accounts
WHERE id == account_id
```

File: src/tgbot_expenses/services/category_service.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+    :type monthly limit: Decimal
+    :return: None
+    """
+    async with AsyncSessionWithEnter(database.engine) as session:
+        category = Category(name=category_name,
+                            monthly_limit=monthly_limit)
+        session.add(category)
+        await session.commit()
+
+    return
```
Comment (atsanda) [inline src/tgbot_expenses/services/category_service.py]: no need for return

File: src/tgbot_expenses/services/category_service.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+    Retrieve the monthly limit for the given category from
+    the 'categories' table.
+
+    :param category_name: The name of the category to retrieve
+                            the limit amount for.
+    :type category_name: str
+    :return: The monthly limit for the category.
+    """
+    async with AsyncSessionWithEnter(database.engine) as session:
+        category_obj = await session.execute(select(Category).where(
```
Comment (atsanda) [inline src/tgbot_expenses/services/category_service.py]: Make specific as with accounts

File: src/tgbot_expenses/services/category_service.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+            Category.name == category_name
+        ))
+        category = category_obj.scalars().first()
+        category.monthly_limit = new_limit
+        await session.commit()
+
+    return
```
Comment (atsanda) [inline src/tgbot_expenses/services/category_service.py]: no need for return

File: src/tgbot_expenses/services/income_service.py [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+        account_obj = await session.execute(select(Account).where(
+            Account.name == account_name
+        ))
+        account = account_obj.scalars().first()
+        income = Income(amount=amount, account_id=account.id)
+        session.add(income)
+        account.balance = account.balance + amount
+        await session.commit()
+
+    return
```
Comment (atsanda) [inline src/tgbot_expenses/services/income_service.py]: remove return

File: src/tgbot_expenses/services/account_service.py [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+                    Account.name == account_from
+                ).with_for_update()
+            )
+            account_obj_to = await session.execute(
+                select(Account).where(
+                    Account.name == account_to
+                ).with_for_update()
+            )
+
+            account_from = account_obj_from.scalars().first()
```
Comment (atsanda) [inline src/tgbot_expenses/services/account_service.py]: Will your bot crash if there is no such an account? If provided id is incorrect?

**Review by atsanda — CHANGES_REQUESTED** [review]

File: src/tgbot_expenses/dialogs/callbacks/settings/limit_settings/changing_limit.py (hunk 14 above)
Comment (Evgenia789) [reply src/tgbot_expenses/dialogs/callbacks/settings/limit_settings/changing_limit.py]: I changed it.

File: src/tgbot_expenses/services/account_service.py (hunk 15 above)
Comment (Evgenia789) [reply src/tgbot_expenses/services/account_service.py]: I changed it.

File: src/tgbot_expenses/services/account_service.py (hunk 16 above)
Comment (Evgenia789) [reply src/tgbot_expenses/services/account_service.py]: I changed it.

File: src/tgbot_expenses/services/category_service.py (hunk 17 above)
Comment (Evgenia789) [reply src/tgbot_expenses/services/category_service.py]: I changed it.

File: src/tgbot_expenses/services/category_service.py (hunk 18 above)
Comment (Evgenia789) [reply src/tgbot_expenses/services/category_service.py]: I changed it.

File: src/tgbot_expenses/services/category_service.py (hunk 19 above)
Comment (Evgenia789) [reply src/tgbot_expenses/services/category_service.py]: I changed it.

File: src/tgbot_expenses/services/income_service.py (hunk 20 above)
Comment (Evgenia789) [reply src/tgbot_expenses/services/income_service.py]: I changed it.

File: src/tgbot_expenses/services/account_service.py (hunk 21 above)
Comment (Evgenia789) [reply src/tgbot_expenses/services/account_service.py]: Added the 'try`-`except` block

File: src/tgbot_expenses/dialogs/callbacks/settings/limit_settings/changing_limit.py (hunk 14 above)
Comment (atsanda) [reply src/tgbot_expenses/dialogs/callbacks/settings/limit_settings/changing_limit.py]: resolved

**Review by atsanda — APPROVED** [review]
ready to merge
there are still things to improve, but overall the requested feature is imlpemented

**Review by atsanda — APPROVED** [review]
